src/lib/parts.ts

```diff
@@ -319,9 +323,12 @@ export class NodePart implements Part {
   }
 
   private _commitPromise(value: Promise<any>): void {
-    this.value = value;
+    this._promise = value;
     value.then((v: any) => {
-      if (this.value === value) {
+      if (this._promise === value) {
+         // This is not strictly necessary, but avoids additional commits
+         // if the value is `noChange`.
```

Comment:
On a re-render after the promise resolves, won't this keep us from checking the Promise identity and make us always call .then()? It seems like we want to only call .then() if the identity of the Promise has changed - treating it almost like a primitive.

Reply:
After the promise resolves, a regular call to `commit(<value from promise>)` will always clear out `this._promise`. That line is there to make sure this also happens when `<value from promise>` is `noChange`, because that is the only possible value skips the reset and returns early (other than a directive).

We have to reset `this._promise` when we render another value, to avoid the following problem sequence:

1. User commits Promise A -> nothing renders
2. User commits some other value -> other value renders
3. Promise A resolves -> result of Promise A renders

Essentially, if we don't reset `this._promise`, unresolved promises will trigger renders after the user rendered something else, which is clearly not desirable.

Because Promises no longer affect `this.value`, we can freely call `then()` on the promise and commit the result, because if the promise already resolved and the value does not change, then the commit will also be cached.

I looked into other ways of solving the problem, but this is the one of the few that avoids all the race conditions that can happen (there are many) between different renders and promises resolving. It's also easily the least complex solution.

One optimisation that can be made is an additional `if (this._promise !== value)` before `value.then`, which ensures that we only queue one `then` if we render multiple times before the promise resolves, but the current construction makes sure that any additional queued `then`s are aborted immediately with `if (this._promise === value)` because the first `then` to resolve will reset `this._promise`. On IE11 where you need Polyfills for Promises, this additional check will probably be beneficial, but on platforms that support Promises it costs one additional equality check, plus the extra bytes.

Reply:
After the promise resolves, a regular call to commit(<value from promise>) will always clear out this._promise. That line is there to make sure this also happens when <value from promise> is noChange, because that is the only possible value skips the reset and returns early (other than a directive).

We have to reset this._promise when we render another value, to avoid the following problem sequence:

User commits Promise A -> nothing renders
User commits some other value -> other value renders
Promise A resolves -> result of Promise A renders
Essentially, if we don't reset this._promise, unresolved promises will trigger renders after the user rendered something else, which is clearly not desirable.

Because Promises no longer affect this.value, we can freely call then() on the promise and commit the result, because if the promise already resolved and the value does not change, then the commit will also be cached.

I looked into other ways of solving the problem, but this is the one of the few that avoids all the race conditions that can happen (there are many) between different renders and promises resolving. It's also easily the least complex solution.

One optimisation that can be made is an additional if (this._promise !== value) before value.then, which ensures that we only queue one then if we render multiple times before the promise resolves, but the current construction makes sure that any additional queued thens are aborted immediately with if (this._promise === value) because the first then to resolve will reset this._promise. On IE11 where you need Polyfills for Promises, this additional check will probably be beneficial, but on platforms that support Promises it costs one additional equality check, plus the extra bytes.